### PR TITLE
feat: support inject and collect barrier from partial graph

### PR DESCRIPTION
**Pull request by wenym1**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Currently on CN, barriers are injected and collected together on a global streaming graph. When supporting partial checkpoint, we may need to support injecting and collecting barriers independently from different partial graphs.

We currently have `ManagedBarrierState` to manage the barrier injected and collected for the global streaming graph. To support the barrier management of different partial graphs, we can reuse this `ManagedBarrierState` and assign one to each partial graph. In this PR, this `ManagedBarrierState` is renamed to `PartialGraphManagedBarrierState`. The original `ManagedBarrierState` will change to hold the state of all partial graphs as `HashMap<PartialGraphId, PartialGraphManagedBarrierState>`. In later partial checkpoint based backfill implementation https://github.com/risingwavelabs/risingwave/pull/17735, the `PartialGraphId` of the global streaming graph is `u32::MAX`, and the one of each creating streaming job is the `table_id` of the `TableFragments`.

Actors can be firstly in the partial graph of a creating streaming job, and then gets merged to the global partial graph when the streaming job is created, which means the partial graph that an actor belongs to is not fixed. Therefore, in this PR we introduce a struct `InflightActorState`, which is an extension to the previous `ActorMutationSubscribers`, and add a field `inflight_barriers: BTreeMap<u64, (PartialGraphId, Option<Arc<Mutation>>)>` to store the inflight barrier of the actor, and also the partial graph id that the actor belongs to in different epochs.

Previously the `InjectBarrierRequest` in the `StreamingControlStream` rpc carries the barrier collection information of the whole streaming graph. In this PR, we add a new field `partial_graph_id` in the `InjectBarrierRequest`, and the original barrier collection information will be carried per partial graph. This PR is just a refactor PR. On meta node global barrier manager, things are still managed on a single global streaming graph, and therefore the `partial_graph_id` is always `u32::MAX`.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by kwannoel** [comment]
You can add a PR description of the conceptual and implementation design, and ping me when it's ready for review. Seems like there's still changes happening.

**Comment by wenym1** [comment]
> You can add a PR description of the conceptual and implementation design, and ping me when it's ready for review. Seems like there's still changes happening.

@kwannoel @BugenZhao @yezizp2012 The PR is ready for review now. I just updated the PR description to add some implementation-wise descriptions. PTAL.

File: proto/stream_service.proto
Comment (kwannoel) [inline proto/stream_service.proto:64]: These are internal APIs right? I think we don't need to care about backwards compat here.

File: proto/stream_service.proto
Comment (kwannoel) [inline proto/stream_service.proto:112]: What is `RemovePartialGraphRequest` for?

File: src/meta/src/barrier/mod.rs
Comment (kwannoel) [inline src/meta/src/barrier/mod.rs:701]: What errors can hit this `warn` branch?

File: proto/stream_service.proto
Comment (wenym1) [reply proto/stream_service.proto:112]: It's for cleaning the stale partial graph. I used to try to let CN clean the empty partial graph by itself, but to handle some corner cases, we may need more complicated actor state management in some previous [commit](https://github.com/risingwavelabs/risingwave/pull/17758/commits/6dbb67fe426dd189d813dbaa9b74458047eff7df). To make this PR simple, I finally chose to let meta explicitly send this request to clean the stale partial graph.

File: src/meta/src/barrier/mod.rs
Comment (wenym1) [reply src/meta/src/barrier/mod.rs:701]: When the CN gets disconnected by accidence but there is no actor running on it.

**Comment by wenym1** [comment]
@BugenZhao Some TODOs left in https://github.com/risingwavelabs/risingwave/pull/17662 are addressed in this PR after merging the latest main branch.

File: src/meta/src/barrier/mod.rs
Comment (hzxa21) [inline src/meta/src/barrier/mod.rs:705]: I don't think we need to return a bool for `self.checkpoint_control.on_collect_failure`. Simply checking `self.state.inflight_actor_infos.actor_map` seems to be sufficient if the purpose here is to detect whether there is any actor allocated for a worker.

File: src/stream/src/task/barrier_manager/progress.rs
Comment (hzxa21) [inline src/stream/src/task/barrier_manager/progress.rs:53]: Is it possible to have `None` returned for any one of this check? Should we have a log in such cases?

File: src/stream/src/task/barrier_manager.rs
Comment (hzxa21) [inline src/stream/src/task/barrier_manager.rs:727]: nits: this can be a varaible with eager initialization instead of a closure given that `all_to_collect()` is used multiple times.

File: src/stream/src/task/barrier_manager.rs
Comment (hzxa21) [inline src/stream/src/task/barrier_manager.rs:855]: Why do we skip reporting the error and reset the stream on `!actor_state.inflight_barriers.is_empty()`? Will we sightly ignore the actor error if the error happens after collecting the previous barrier but before issuing the current barrier?

File: src/meta/src/barrier/mod.rs
Comment (wenym1) [reply src/meta/src/barrier/mod.rs:705]: The `self.state.inflight_actor_infos.actor_map` stores the latest map that has been applied with mutation all the injected barrier. Therefore, for example, if we have a drop mview command that drops all mviews, after its barrier is injected and wait for collection, the `self.state.inflight_actor_infos.actor_map` is empty, and only checking it will ignore this error, but we should not ignore it.

File: src/stream/src/task/barrier_manager.rs
Comment (wenym1) [reply src/stream/src/task/barrier_manager.rs]: The actors exited with error will be saved in the `failure_actors`, and if the later `inject_barrier` has to collect the barrier from actor in `failure_actors`, we will return with error. 

If we won't check `!actor_state.inflight_barriers.is_empty()` and return the error anyway, we may trigger unnecessary recover for error from actors that are temporarily created in create mview command but has not added to the streaming graph yet.

I just change the current `is_stopping` bool field in `InflightActorState` to a `InflightActorStatus` enum with variants `NotStarted`, `Running` and `Stopping`. We will report the error when `!actor_state.inflight_barriers.is_empty()` or the status is `Running`.

**Comment by wenym1** [comment]
In later development, I found that we don't have scenarios to inject barrier to multiple partial graph in a single request. Therefore, I remove the newly introduced `PartialGraphInfo`, and put the fields back to `InjectBarrierRequest` and add a new field `partial_graph_id` to it.

File: src/stream/src/task/barrier_manager.rs
Comment (hzxa21) [reply src/stream/src/task/barrier_manager.rs]: > We will report the error when !actor_state.inflight_barriers.is_empty() or the status is Running.

I think you mean when `!actor_state.inflight_barriers.is_empty()` or the status is **not** Running?

File: proto/stream_service.proto
Comment (kwannoel) [reply proto/stream_service.proto:107]: > clean the empty partial graph by itself

What do you mean by "empty" partial graph?

File: src/stream/src/task/barrier_manager.rs
Comment (wenym1) [reply src/stream/src/task/barrier_manager.rs]: > > We will report the error when !actor_state.inflight_barriers.is_empty() or the status is Running.
> 
> I think you mean when `!actor_state.inflight_barriers.is_empty()` or the status is **not** Running?

No, I meant `Running`. 

Actor can be of 4 status: 
1. spawned but not added to the streaming graph, which is the `InflightActorStatus::NotStarted`
2. added to the streaming graph, and there are some inflight barriers not collected to it yet, which is `InflightActorStatus::Running`, with non-empty `actor_state.inflight_barriers`
3. added to the streaming graph, and all barriers on it has been collected, no inflight barrier, waiting for injecting next barrier, which is the `InflightActorStatus::Running` with empty `actor_state.inflight_barriers`
4. added to the streaming graph, issued with a stop barrier, waiting for some barriers to be collected, which is `InflightActorStatus::Stopping`

Before this PR, and also at the commit that this comment was on, when we see an actor failure, we ignore it on status 1, report error immediately on status 2 and 4, report error on next inject barrier on status 3.

IIUC, the expected behavior should be, ignore the error on status 1, and report error immediately on status 2, 3 and 4. This difference can be distinguished by `!actor_state.inflight_barriers.is_empty() or the status is Running`. When `!actor_state.inflight_barriers.is_empty()`, the actor must be at status 2 or 4. When the actor has empty inflight barriers, but at `InflightActorStatus::Running`, the actor must be at status 3.

File: proto/stream_service.proto
Comment (wenym1) [reply proto/stream_service.proto:107]: In the partial checkpoint based backfill implementation, each creating streaming job will have a separate partial graph. When streaming jobs finish creating, it will be merged to the original global streaming graph, and then the original partial graph will become empty and needs to be clear.

File: proto/stream_service.proto
Comment (kwannoel) [inline proto/stream_service.proto:84]: What's the partial graph id referring to here?

Comment (kwannoel) [reply proto/stream_service.proto:84]: Ah I see, it's the id of the partial graph where the barrier is collected from.

File: src/meta/src/barrier/info.rs
Comment (kwannoel) [inline src/meta/src/barrier/info.rs:27]: Why remove `CommandNewFragmentInfo`?

File: proto/stream_service.proto
Comment (wenym1) [reply proto/stream_service.proto:84]: Barrier are injected and collected per partial graph. This refers to the partial graph that has finish collecting the barrier.

File: src/meta/src/manager/catalog/fragment.rs
Comment (kwannoel) [inline src/meta/src/manager/catalog/fragment.rs:177]: Could we add some docs about the lifetime and usage of this struct?

Like when it is constructed, what it is used for, when it is dropped?

File: src/meta/src/barrier/info.rs
Comment (wenym1) [reply src/meta/src/barrier/info.rs:27]: The `CommandNewFragmentInfo` is exactly the same as `InflightFragmentInfo`, so using `InflightFragmentInfo` is enough.

File: src/meta/src/barrier/info.rs
Comment (kwannoel) [inline src/meta/src/barrier/info.rs:104]: Why do we no longer need this?

File: src/meta/src/barrier/info.rs
Comment (kwannoel) [inline src/meta/src/barrier/info.rs:158]: Could you elaborate a bit more on `is_injectable`, I see its use removed in various places, but I don't know what it was used for previously, and why it's obsolete now.

File: src/stream/src/task/barrier_manager.rs
Comment (kwannoel) [inline src/stream/src/task/barrier_manager.rs:570]: I suppose there's no notion of a global `current_epoch` anymore right? Hence the change. Because there could be mulitple "current_epoch" , one for each partial graph.

File: src/meta/src/manager/catalog/fragment.rs
Comment (wenym1) [reply src/meta/src/manager/catalog/fragment.rs:177]: The `InflightFragmentInfo` is just a plain info rather than some states, so it does not have any lifetime transition. Do you mean `InflightActorState`?

File: src/meta/src/barrier/info.rs
Comment (wenym1) [reply src/meta/src/barrier/info.rs:104]: We no longer maintain the `actor_map_to_send` in the `InflightActorInfo`. When `inject_barrier`, we will calculate the `actor_ids_to_send` by the providing fragment info.

File: src/meta/src/barrier/info.rs
Comment (wenym1) [reply src/meta/src/barrier/info.rs:158]: The `is_injectable` was to maintaining the `actor_map_to_send`, which is removed in this PR.

The `actor_map_to_send` means the actor that directly receives the barriers from barrier sender on `LocalBarrierWorker` instead of receiving from upstream.

File: src/stream/src/task/barrier_manager.rs
Comment (wenym1) [reply src/stream/src/task/barrier_manager.rs:570]: I think it's not. For a single actor, a `current_epoch` only exists in partial graph. In other word, it's not possible for an actor to exist in multiple partial graphs at a same `current_epoch`.

**Comment by kwannoel** [comment]
> In later partial checkpoint based backfill implementation https://github.com/risingwavelabs/risingwave/pull/17735, the PartialGraphId of the global streaming graph is u32::MAX, and the one of each creating streaming job is the table_id of the TableFragments.

Why is there still a "global streaming graph", I thought after the partial ckpt based implementation there shouldn't be anymore global graph, and instead just a set of partial graphs?

**Comment by wenym1** [comment]
> > In later partial checkpoint based backfill implementation #17735, the PartialGraphId of the global streaming graph is u32::MAX, and the one of each creating streaming job is the table_id of the TableFragments.
> 
> Why is there still a "global streaming graph", I thought after the partial ckpt based implementation there shouldn't be anymore global graph, and instead just a set of partial graphs?

Since in frontend node we still manage a global snapshot, all the created streaming graphs still shares a single global streaming graph, and barriers are injected and collected together at this global streaming graph. The creating streaming job will do the backfill in an individual partial graph, and after finish creating, it will join the global streaming graph of the collected streaming jobs. Therefore, we still need this global streaming graph to maintain the global snapshot.

File: src/meta/src/barrier/recovery.rs
Comment (kwannoel) [inline src/meta/src/barrier/recovery.rs:393]: Why we pass the same info twice

File: src/meta/src/stream/stream_manager.rs
Comment (kwannoel) [inline src/meta/src/stream/stream_manager.rs:1034]: What's this refactor for?

File: src/stream/src/task/barrier_manager.rs
Comment (hzxa21) [reply src/stream/src/task/barrier_manager.rs]: > IIUC, the expected behavior should be, ignore the error on status 1, and report error immediately on status 2, 3 and 4. This difference can be distinguished by !actor_state.inflight_barriers.is_empty() or the status is Running. When !actor_state.inflight_barriers.is_empty(), the actor must be at status 2 or 4. When the actor has empty inflight barriers, but at InflightActorStatus::Running, the actor must be at status 3.

Agree. Discussed offline. Previously we handled case 3 incorrectly in the codes and it is fixed in the latest commit.

File: src/storage/src/hummock/event_handler/uploader/mod.rs
Comment (kwannoel) [inline src/storage/src/hummock/event_handler/uploader/mod.rs:872]: What's this change here in relation to?

**Review by hzxa21 — APPROVED** [review]
LGTM.

File: src/meta/src/stream/stream_manager.rs
Comment (wenym1) [reply src/meta/src/stream/stream_manager.rs:1034]: After this PR, the `FakeStreamService` should take the `worker_node.id` before spawned. Therefore, we changed to create and spawn it after `add_worker_node` to get the `worker_node.id`.

**Review by kwannoel — APPROVED** [review]
Rest LGTM. @yezizp2012 could you PTAL as well?

File: src/storage/src/hummock/event_handler/uploader/mod.rs
Comment (wenym1) [reply src/storage/src/hummock/event_handler/uploader/mod.rs:872]: The previous assertion was hit in some previous commits and get fixed in a later commit. 

This assertion was a insignificant sanity check, and I am concerned it might be too strict in production env and cause unnecessary panic, so I changed to having a warning log.

File: src/meta/src/barrier/recovery.rs
Comment (wenym1) [reply src/meta/src/barrier/recovery.rs:393]: The two parameters have different semantics. It just happens to be the same one in recovery.

**Comment by BugenZhao** [comment]
I would also like to take a look later. 🙏

**Comment by wenym1** [comment]
> I would also like to take a look later. 🙏

Will merge the PR first to unblock further development. You can leave some comments on the PR and I will address it in later PRs.

**Comment by yezizp2012** [comment]
> Rest LGTM. @yezizp2012 could you PTAL as well?

Yes sure. I will take a look tmr. 🫡

**Pull request merged**
